fw/cache.c

```diff
@@ -2824,6 +2857,20 @@ tfw_cache_build_resp(TfwHttpReq *req, TfwCacheEntry *ce, long lifetime,
 	}
 	resp->content_length = ce->body_len;
 
+	if (has_trailers) {
+		struct sk_buff *skb, *nskb = nskb_head;
+		resp->req->stream->xmit.t_len = t_len;
+		skb_set_tfw_flags(nskb, SS_F_HTTP2_FRAME_START);
+		do {
+			skb_set_tfw_cb(nskb, stream_id);
+			skb_set_tfw_flags(nskb, SS_F_HTTT2_FRAME_TRAILER_HEADERS);
+			skb = nskb->next;
+			nskb->next = nskb->prev = NULL;
+			ss_skb_queue_tail(&resp->msg.skb_head, nskb);
+			nskb = skb;
+		} while (nskb != nskb_head);
```

Comment:
yes, I agree we should handle it in a normal way instead of hacking. As I said, that was one of my two options when I was working on it.
but I need to emphasize that under the current code framework, this hacking way is already the most efficient way to achieve the goal. **The extra conditions exist because the framework requires to loop over the whole header list including trailers so that it gets the correct pointer to the body, if you remove the inner conditions, then you need two loops, which is worse because each loop needs to iterate the whole header list (the cache entry has one header list, which didn't distinguish the regular headers and trailers).**

Reply:
Let me repeat my opinion. The hacking way works, no matter what disadvantage you said (I also admit). But why did I choose it? It's just one week of development work, but for the normal way, maybe several weeks and more extra time to ensure it does not bring in new bugs because the amount of code will be much larger and involves more aspects, which is normally more error-prone. Not to mention, as @const-t said, it depends on https://github.com/tempesta-tech/tempesta/pull/1973 too, I need to cooperate with that PR including the framing redesign.

Reply:
All in all, shall I start the refactoring later? or after https://github.com/tempesta-tech/tempesta/pull/1973?